<launch>
  <arg name="rgb_camera_info_url"   default="" />
  <arg name="depth_camera_info_url" default="" />
  <group>       <!--group only for remapping-->
    <remap from="/camera_remote/depth/image_raw" 
           to="/camera/depth/image_raw"/>
    <remap from="/camera_remote/depth/camera_info" 
           to="/camera/depth/camera_info"/>
    <remap from="/camera_remote/depth/image_rect_raw"
           to="/camera/depth/image_rect_raw"/>

    <remap from="/camera_remote/depth_registered/image_raw"
           to="/camera/depth_registered/image_raw"/>
    <remap from="/camera_remote/rgb/image_raw"
           to="/camera/rgb/image_raw"/>
    <remap from="/camera_remote/rgb/camera_info"
           to="/camera/rgb/camera_info"/>
    <remap from="/camera_remote/depth_registered/hw_registered/image_rect_raw"
           to="/camera/depth_registered/hw_registered/image_rect_raw"/>
    <include file="$(find openni2_launch)/launch/openni2.launch">


      <arg name="publish_tf" value="false" />
      <arg name="camera" value="camera_remote" />
      <arg name="depth_registration" value="true" />
      <arg name="load_driver" value="false" />
      <arg name="rgb_processing" value="true" />
      <arg name="depth_processing" value="true" />
      <arg name="depth_registered_processing" value="true" />
      <arg name="rgb_camera_info_url" value="$(arg rgb_camera_info_url)" />
      <arg name="depth_camera_info_url" value="$(arg depth_camera_info_url)" />
    </include>
  </group>
  
  <group ns="camera_remote">
    <!--                      -->
    <!-- camera_info relaying -->
    <!--                      -->
    <group ns="rgb">
      <node pkg="topic_tools"
            type="relay"
            name="relay_camera_info"
            args="/camera/rgb/camera_info /camera_remote/rgb/camera_info" />
    </group>
    <group ns="depth">
      <node pkg="topic_tools"
            type="relay"
            name="relay_camera_info"
            args="/camera/depth/camera_info /camera_remote/depth/camera_info" />
    </group>
    <!--                       -->
    <!-- downsample pointcloud -->
    <!--                       -->
    <node pkg="nodelet"
          type="nodelet"
          name="depth_registered_half"
          args="load jsk_pcl/ResizePointsPublisher camera_remote_nodelet_manager">
      <remap from="~input" to="depth_registered/points" />
      <remap from="~output" to="~points" />
      <param name="step_x" value="2" />
      <param name="step_y" value="2" />
    </node>
    <node pkg="nodelet"
          type="nodelet"
          name="depth_half"
          args="load jsk_pcl/ResizePointsPublisher camera_remote_nodelet_manager">
      <remap from="~input" to="depth/points" />
      <remap from="~output" to="~points" />
      <param name="not_use_rgb" value="true" />
      <param name="step_x" value="2" />
      <param name="step_y" value="2" />
    </node>
    <node pkg="nodelet"
          type="nodelet"
          name="depth_registered_quater"
          args="load jsk_pcl/ResizePointsPublisher camera_remote_nodelet_manager">
      <remap from="~input" to="depth_registered/points" />
      <remap from="~output" to="~points" />
      <param name="step_x" value="4" />
      <param name="step_y" value="4" />
    </node>
    <node pkg="nodelet"
          type="nodelet"
          name="depth_quater"
          args="load jsk_pcl/ResizePointsPublisher camera_remote_nodelet_manager">
      <remap from="~input" to="depth/points" />
      <remap from="~output" to="~points" />
      <param name="not_use_rgb" value="true" />
      <param name="step_x" value="4" />
      <param name="step_y" value="4" />
    </node>
    <!--            -->
    <!-- resize rgb -->
    <!--            -->
    <node pkg="nodelet"
          type="nodelet"
          name="rgb_half"
          args="load resized_image_transport/ImageResizer camera_remote_nodelet_manager">
      <param name="resize_scale_x" value="0.5" />
      <param name="resize_scale_y" value="0.5" />
      <remap from="~input/image" to="rgb/image_rect_color" />
      <remap from="~output/image" to="~image_rect_color" />
    </node>
    <node pkg="nodelet"
          type="nodelet"
          name="rgb_quater"
          args="load resized_image_transport/ImageResizer camera_remote_nodelet_manager">
      <param name="resize_scale_x" value="0.25" />
      <param name="resize_scale_y" value="0.25" />
      <remap from="~input/image" to="rgb/image_rect_color" />
      <remap from="~output/image" to="~image_rect_color" />
    </node>
  </group>
  
  
</launch>
